### ISSUE TITLE
Fine-tuning the pre-trained model

### ISSUE THREAD

**Issue by crisostomi** [facebookresearch/audioseal#40]
Hello!

I am trying to fine-tune the pre-trained model over a new dataset, but I am having a hard time finding the checkpoint to provide as a `continue_from` argument to `dora`. The provided checkpoints are in fact already separated into generator and detector, and cannot be used in the training pipeline which expects a single checkpoint for the whole model. 
Am I missing something? 

Thanks a lot!

**Comment by crisostomi** [comment]
Yes, I managed to load the pretrained generator and detector for inference, but I need the single non-preprocessed checkpoint to be able to use it as initialization when fine-tuning it with `dora run continue_from <PATH_TO_CKPT>`.

**Comment by crisostomi** [comment]
In case anyone might have the same problem, I quickly wrote a helper function to combine the checkpoints back into a single one

```python
def combine(
    generator_checkpoint: str, detector_checkpoint: str, output_checkpoint: str
):
    """Combine generator and detector checkpoints into a single checkpoint."""

    gen_ckpt = torch.load(generator_checkpoint)
    det_ckpt = torch.load(detector_checkpoint)

    combined_ckpt = {
        "xp.cfg": gen_ckpt["xp.cfg"],  # assuming the configs are identical
        "model": {},
    }

    # Add generator layers with appropriate prefix
    for layer in gen_ckpt["model"].keys():
        new_layer = f"generator.{layer}"
        combined_ckpt["model"][new_layer] = gen_ckpt["model"][layer]

    # Add detector layers with appropriate prefix
    for layer in det_ckpt["model"].keys():
        new_layer = f"detector.{layer}"
        combined_ckpt["model"][new_layer] = det_ckpt["model"][layer]

    # Special case for 'msg_processor.msg_processor.weight'
    if "msg_processor.msg_processor.weight" in gen_ckpt["model"]:
        combined_ckpt["model"]["msg_processor.msg_processor.0.weight"] = gen_ckpt[
            "model"
        ]["msg_processor.msg_processor.weight"]

    torch.save(combined_ckpt, output_checkpoint)
```

**Comment by antoine-tran** [comment]
thanks, do you want to open a PR for this ?

**Comment by antoine-tran** [comment]
btw, `continue_from` is not a real fine-tuning recipe but rather to continue the long-running / checkpointed model. It performs full updates of the model parameters and can lead to catastrophic forgetting.

**Comment by crisostomi** [comment]
> thanks, do you want to open a PR for this ?

Sure, will do!

> btw, continue_from is not a real fine-tuning recipe but rather to continue the long-running / checkpointed model. It performs full updates of the model parameters and can lead to catastrophic forgetting.

Thanks, I see, do you have any hints how you would do a fine-tuning instead?

**Comment by hadyelsahar** [comment]
@crisostomi  I know loading ckpts works to resume training works well and tested. I would say you need to debug here to make sure it does the right thing and your weights are loaded correctly. 

https://github.com/facebookresearch/audiocraft/blob/adf0b04a4452f171970028fcf80f101dd5e26e19/audiocraft/solvers/base.py#L342

> Thanks, I see, do you have any hints how you would do a fine-tuning instead?

Depends on what you want to achieve, if you want to make the detector more robust then you could freeze the generator. 
Also make sure that `dora run ... continue_from` loads the correct parameters for training that you need.

**Comment by crisostomi** [comment]
@hadyelsahar thanks for the answer!

I guess we still miss the weights for the adversarial losses to resume the training from the pre-trained checkpoint, right?

**Comment by pierrefdz** [comment]
Yes indeed. 
Have you tried fine-tuning without the per-loaded weights for the adversarial networks?

**Comment by royrs** [comment]
@crisostomi Did you succeed fine-tuning the model?
I've tried using the function you wrote to combine the weights but I'm having issues.

I'm trying to just evaluate at the moment using:
```dora run solver=watermark/robustness execute_only=evaluate dset=audio/voxpopuli continue_from=combined.pth +dummy_watermarker.nbits=16```

At first I got an error that it can't find "best_state". It seems that it should be identical to model, so I copied it.
However, I get error about non-existing layer
```    
self.states[name][k].copy_(v)
KeyError: 'detector.detector.0.model.0.conv.conv.weight_g'
```

I tried to run a training from scratch and compare the the layers from the checkpoint and the layers of the combined checkpoint and it seems there are about ~28 different layers.

Is there some config file I'm missing that I need to change to use the weights given in this repo?
